### ISSUE TITLE
match goal fails to match primitive projections after reduction

### ISSUE THREAD

**Issue by jwiegley** [coq/coq#6994]
#### Version

8.7.2

#### Operating system

macOS 10.13.3

#### Description of the problem

The attached code shows a case where after changing the goal and using `vm_compute`, I cannot use `match goal` to match exactly what is in the goal.

I was unable to reproduce this with a toy example, so I've reduced my full problem down to this attachment.

[Bug.zip](https://github.com/coq/coq/files/1812662/Bug.zip)

**Comment by jwiegley** [comment]
The purpose behind this code is to provide a reflection-based rewriting tactics, so that I can rewrite categorical compositions regardless of associativity.  The impact of the reported bug is that I only get to use the tactic exactly once.

**Comment by JasonGross** [comment]
This is an issue with primitive projections.  Here's a reduced example:
```coq
Set Primitive Projections.
Class Setoid A := { equiv : A -> A -> Type }.

Goal forall A B C D, @equiv A B C D.
  intros.
  lazymatch goal with
  | [ |- @equiv _ _ _ _ ] =>
    idtac
  end. (* success *)
  vm_compute.
  lazymatch goal with
  | [ |- @equiv _ _ _ _ ] =>
    idtac
  end. (* fails *)
```

**Comment by jwiegley** [comment]
@JasonGross Thank you!

**Comment by RalfJung** [comment]
I think I just ran into this. I wasted an hour debugging why `eauto` with our custom hints would not solve a goal, and it boils down to `match goal` not matching a pattern with a goal when the pattern definitely applies. The term that fails to match is a primitive projection. I was unable to reproduce this in any smaller example.

**Comment by mattam82** [comment]
This is due to the dreaded `unfolded` status of primitive projections that was introuced for compatibility with non-primitive ones (if you unset primitive projections, you'll see that indeed the second match goal cannot succeed).

**Comment by ppedrot** [comment]
> This is due to the dreaded unfolded status of primitive projections

Pretty please, let's get rid of that wart a soon as possible and the devil with the *borken* compat layer.

**Comment by RalfJung** [comment]
Yes please indeed. Can we throw MPI consortium resources at this?^^

**Comment by math-fehr** [comment]
I also ran into this, not after doing a reduction, but when matching on a record that has primitive projections:

```Coq
Local Set Primitive Projections.
Record r {T: Type} := constr { arg1: T; arg2: T }.
Arguments r : clear implicits.
Arguments constr {T} arg1 arg2 : assert.

Definition foo (x: r bool) :=
  match x with
  | constr true _ => true
  | constr _ _ => false
  end.

Goal forall x, foo x = false -> True.
  intros.
  unfold foo in *.
  (* This match fails *)
  Fail match goal with
       | [ H: (if arg1 _ then _ else _) = _ |- _ ] => idtac H
       end.
Abort.
```

**Comment by ppedrot** [comment]
Fixed by #15559.
